### ISSUE TITLE
Add error handlers to http file server

### ISSUE THREAD

**Issue by nhooyr** [golang/go#27139]
Noticed that in Go 1.11 a error handler field was added to httputil's ReverseProxy to allow sending allow writing a custom error.

I think the same should be done for http.FileServer, http.ServeFile and http.ServeContent.

I don't see an easy way to do this without breaking the API or adding more API so maybe this is better for Go 2.

**Comment by nhooyr** [comment]
This would be really janky but one way to do it would be to set the error handler within the request's context, I'm not sure if we want to go down this path though, seems like a misuse of the context.

**Comment by Azareal** [comment]
You can do it by wrapping the http.ResponseWriter with your own writer and intercepting the call to WriteHeader, and then, using the status code from there to decide whether you want to serve an error. It is a little slow and involved though.

**Comment by Azareal** [comment]
Actually, scratch that, I'm not sure that actually works.

**Comment by nhooyr** [comment]
Yup. The issue with that is implementing all of the response writer upgrade interfaces afterwards. Super janky especially when the underlying response writer may not support all of them.
